src/main/java/net/neoforged/moddevgradle/internal/RunUtils.java

```diff
@@ -231,69 +219,25 @@ private static Project findSourceSetProject(Project someProject, SourceSet sourc
-                                                                  Provider<Set<ModModel>> modsProvider,
-                                                                  boolean includeUnitTests) {
-        var folders = buildModFolders(project, modsProvider, includeUnitTests, (sourceSet, output) -> {
-            output.from(findSourceSetProject(project, sourceSet).getProjectDir().toPath()
-                    .resolve("bin")
-                    .resolve(sourceSet.getName()));
-        });
-
-        var modFoldersProvider = project.getObjects().newInstance(ModFoldersProvider.class);
-        modFoldersProvider.getModFolders().set(folders);
-        return modFoldersProvider;
-    }
-
-    private static String getIdeaOutName(final SourceSet sourceSet) {
-        return sourceSet.getName().equals(SourceSet.MAIN_SOURCE_SET_NAME) ? "production" : sourceSet.getName();
-    }
-
-    private static Provider<Map<String, ModFolder>> getModFoldersForGradle(Project project, Provider<Set<ModModel>> modsProvider, boolean includeUnitTests) {
-        return buildModFolders(project, modsProvider, includeUnitTests, (sourceSet, output) -> {
+    public static Provider<Map<String, ModFolder>> getModFoldersForGradle(Project project,
+                                                                          Provider<Set<ModModel>> modsProvider,
+                                                                          @Nullable Provider<ModModel> testedMod) {
+        return buildModFolders(project, modsProvider, testedMod, (sourceSet, output) -> {
             output.from(sourceSet.getOutput());
         });
     }
 
-    private static Provider<Map<String, ModFolder>> buildModFolders(Project project, Provider<Set<ModModel>> modsProvider, boolean includeUnitTests, BiConsumer<SourceSet, ConfigurableFileCollection> outputFolderResolver) {
-        var extension = ExtensionUtils.findExtension(project, NeoForgeExtension.NAME, NeoForgeExtension.class);
-        var testedModProvider = extension.getUnitTest().getTestedMod()
-                .filter(m -> includeUnitTests)
-                .map(Optional::of)
-                .orElse(Optional.empty());
+    public static Provider<Map<String, ModFolder>> buildModFolders(Project project,
+                                                                   Provider<Set<ModModel>> modsProvider,
+                                                                   @Nullable Provider<ModModel> testedModProvider,
+                                                                   BiConsumer<SourceSet, ConfigurableFileCollection> outputFolderResolver) {
+        // Convert it to optional to ensure zip will be called even if no mod under test is present.
+        if (testedModProvider == null) {
+            testedModProvider = project.provider(() -> null);
+        }
+        var optionalTestedModProvider = testedModProvider.map(Optional::of).orElse(Optional.empty());
```

Comment:
Should be `ofNullable`, no? Or does the provider returning `null` make it count as "unset"?